Add genesis config for nft

pmensik opened the pull request:
The initial idea for a genesis config of the NFT pallet. The config accepts `BTreeMap<T::AccountId, BTreeMap<(Vec<u8>, T::ClassData), Vec<(Vec<u8>, T::TokenData)>>>` that basically says each account has a map where the key is a vector of tuples for token class data (metadata and ClassData) with the value of vector of tokens to be minted. 
However, I can't get this code built, so I will appreciate any help with finishing this PR

```
   Compiling orml-nft v0.3.5-dev (/home/pmensik/work/polkadot/open-runtime-module-library/nft)
error[E0463]: can't find crate for `serde`
   --> src/lib.rs:100:1
    |
100 | / decl_storage! {
101 | |     trait Store for Module<T: Config> as NonFungibleToken {
102 | |         /// Next available class ID.
103 | |         pub NextClassId get(fn next_class_id): T::ClassId;
...   |
161 | |     }
162 | | }
    | |_^ can't find crate
    |
```

xlc commented on nft/Cargo.toml:
```diff
@@ -9,15 +9,15 @@ edition = "2018"
 codec = { package = "parity-scale-codec", version = "1.3.0", default-features = false }
-sp-std = { version = "2.0.0", default-features = false }
-sp-runtime = { version = "2.0.0", default-features = false }
+sp-std = {  version = "2.0.0", default-features = false, git = "https://github.com/paritytech/substrate", rev = "b2750359" }
```
please revert chose change and update Cargo.dev.toml instead.

I've already bumped Substrate version at #350 so no need to bump again

xlc commented on nft/src/lib.rs:
```diff
@@ -91,6 +91,12 @@ pub type ClassInfoOf<T> =
 pub type TokenInfoOf<T> = TokenInfo<<T as frame_system::Config>::AccountId, <T as Config>::TokenData>;
 
+#[derive(Encode, Decode, Clone, Eq, PartialEq, RuntimeDebug)]
+pub struct NftGenesisConfig<AccountId, ClassData, TokenData> {
```
this isn't used at all?

xlc commented on nft/src/lib.rs:
```diff
@@ -109,6 +115,50 @@ decl_storage! {
 		pub TokensByOwner get(fn tokens_by_owner): double_map hasher(twox_64_concat) T::AccountId, hasher(twox_64_concat) (T::ClassId, T::TokenId) => Option<()>;
 	}
+	add_extra_genesis {
+		config(endowed_accounts): BTreeMap<T::AccountId, BTreeMap<(Vec<u8>, T::ClassData), Vec<(Vec<u8>, T::TokenData)>>>;
```
this type is super confusing... you can use type alias to improve it. something like

```
type GenesisTokenData<T> = (Vec<u8>, <T as Trait>::TokenData);
type GenesisClassData<T> = (Vec<u8>, <T as Trait>::ClassData);)
```

Also instead of `AccountId => (ClassData => TokenData)`, I will prefer `ClassData => (AccountId, TokenData)`

xlc commented:
You need those to fix the serde error

https://github.com/open-web3-stack/open-runtime-module-library/blob/2de1e024781a7744a3aeb82229ebcb555163452a/auction/Cargo.toml#L11

https://github.com/open-web3-stack/open-runtime-module-library/blob/2de1e024781a7744a3aeb82229ebcb555163452a/auction/Cargo.toml#L28

pmensik commented:
@xlc I've updated the solution and changed a type according to your comment (I've just added an `AccountId` to a `GenesisClassData` because class data needs an owner as well). Thanks for your review!

xlc commented on nft/Cargo.toml:
```diff
@@ -8,20 +8,22 @@ authors = ["Acala Developers"]
+serde = { version = "1.0.111", optional = true }
 codec = { package = "parity-scale-codec", version = "1.3.0", default-features = false }
-sp-std = { version = "2.0.0", default-features = false }
+sp-std = {  version = "2.0.0", default-features = false }
```
```suggestion
sp-std = { version = "2.0.0", default-features = false }
```

xlc commented on nft/Cargo.toml:
```diff
@@ -8,20 +8,22 @@ authors = ["Acala Developers"]
 
 [dev-dependencies]
-sp-io = { version = "2.0.0", default-features = false }
+sp-io = {  version = "2.0.0", default-features = false }
```
```suggestion
sp-io = { version = "2.0.0", default-features = false }
```

pmensik replied:
Fixed, thanks for the note

xlc commented on nft/src/lib.rs:
```diff
@@ -61,7 +61,7 @@ pub trait Config: frame_system::Config {
 	type TokenId: Parameter + Member + AtLeast32BitUnsigned + Default + Copy;
 	/// The class properties type
-	type ClassData: Parameter + Member;
+	type ClassData: Parameter + Member + Ord;
```
I don't like this additional trait bound just so we can have it in genesis.

One way to avoid it is just use `Vec` instead of `BTreeMap`

pmensik replied:
@xlc Hmm, it's probably not a good idea to restrict the type furthermore. However, I feel that the map is a really good fit in this situation (because it's a mapping of token classes to tokens themselves), is it possible to use another implementation that doesn't require `Ord`? Like this [one](https://substrate.dev/rustdocs/v2.0.0/hashbrown/struct.HashMap.html). On the other hand, introducing an external dependency doesn't sound like a good idea either. So something like this feel alright?

```
// each token will have an owner
pub type GenesisTokenData<T> = (T::AccountId, Vec<u8>, <T as Config>::TokenData);
// each token class will contain its tokens
pub type GenesisTokens<T> = (
	<T as frame_system::Config>::AccountId,
	Vec<u8>,
	<T as Config>::ClassData,
	Vec<GenesisTokenData<T>>
);
config(tokens): Vec<GenesisTokens<T>> {}
```

xlc replied:
Yep that should work.

pmensik replied:
Fixed, I just wonder whether should I add any documentation for the types (in the form comments). Or is self-explanatory enough?

xlc replied:
some simple comment will be helpful, especially for `Vec<u8>` part

pmensik replied:
Done

xlc commented on nft/src/lib.rs:
```diff
@@ -109,6 +116,28 @@ decl_storage! {
+
+		build(|config: &GenesisConfig<T>| {
+			config.endowed_accounts.iter().for_each(|(class_data, tokens)| {
+				let result = Module::<T>::create_class(&class_data.0, class_data.1.to_vec(), class_data.2.clone());
```
```suggestion
				let class_id = Module::<T>::create_class(&class_data.0, class_data.1.to_vec(), class_data.2.clone()).expect("Create class cannot fail while building genesis");
```

Never ignore error silencely

pmensik replied:
Thank you, I didn't think about `expect`

xlc commented:
Please merge master to fix CI

xlc reviewed (approved):
Thanks for the contribution!
